src/lib/agent/MessageReceiver.ts

```diff
@@ -4,37 +4,90 @@ import { Dispatcher } from './Dispatcher';
 import { EnvelopeService } from './EnvelopeService';
 import { UnpackedMessage } from '../types';
 import { MessageType } from '../protocols/routing/messages';
+import { MessageContext } from './models/MessageContext';
+import { ConnectionService } from '../protocols/connections/ConnectionService';
+import { AgentMessage } from './AgentMessage';
+import { MessageTransformer } from './MessageTransformer';
 
 class MessageReceiver {
   config: AgentConfig;
   envelopeService: EnvelopeService;
+  connectionService: ConnectionService;
   dispatcher: Dispatcher;
 
-  constructor(config: AgentConfig, envelopeService: EnvelopeService, dispatcher: Dispatcher) {
+  constructor(
+    config: AgentConfig,
+    envelopeService: EnvelopeService,
+    connectionService: ConnectionService,
+    dispatcher: Dispatcher
+  ) {
     this.config = config;
     this.envelopeService = envelopeService;
+    this.connectionService = connectionService;
     this.dispatcher = dispatcher;
   }
 
-  async receiveMessage(inboundPackedMessage: any) {
+  /**
+   * Receive and handle an inbound DIDComm message. It will unpack the message, transform it
+   * to it's corresponding message class and finaly dispatch it to the dispatcher.
+   *
+   * @param inboundPackedMessage the message to receive and handle
+   */
+  public async receiveMessage(inboundPackedMessage: any) {
     logger.logJson(`Agent ${this.config.label} received message:`, inboundPackedMessage);
-    let inboundMessage;
 
+    const unpackedMessage = await this.unpackMessage(inboundPackedMessage);
+
+    logger.logJson('inboundMessage', unpackedMessage);
+
+    const message = await this.transformMessage(unpackedMessage);
+    let senderKey = unpackedMessage.sender_verkey;
+    let connection = undefined;
+    if (senderKey && unpackedMessage.recipient_verkey) {
+      connection = (await this.connectionService.findByVerkey(unpackedMessage.recipient_verkey)) || undefined;
+
+      // we validate the sender key if it is the one we have the connection with
+      // otherwise everyone could send message for our key, and we would just accept
+      // it as if it was send by the connection.
+      // TODO: does this correctly work for the connection process? Keys can be swapped during the protocol
+      if (connection && connection.theirKey != null && connection.theirKey != senderKey) {
```

Comment:
"Keys can be swapped during the protocol" - Do you mean during the connection protocol, after the Accept Invitation step?

Reply:
If so, we should allow the change just for that case, otherwise, it shouldn't be allowed.

When making a connection, I create some `did`, `verkey` and send it in invitation to other party. I get `theirDid`, `theirVerkey` in connection request and respond with `myDid` and `myVerkey` in connection response.

Then, when I get a message. The `unpack` operation tries to find  `recipient_key` between stored keys in my wallet and decrypt the message with it. But I still need to check if `sender_verkey`, if it’s authCrypt communication, is equal to `connection.theirVerkey` I stored in DIDExchange process. It `sender_verkey` is different, someone who knows `myVerkey` in this DID connection could send me a message signed with his key.

Reply:
Yes I meant during the connection protocol. Will update the comments to make it a bit more clear.

The question on my side if this will work with the current checks:

```
alice: invitation with public did
bob: request with pairwise did
alice: response with pairwise did
```

when bob receives the response with the pairwise did the sender key is different than the one from the invitation. It is then linked together because the attached did document is signed with the same key as the invitation. After verifying this we should update theirkey.

Will look into it this weekend.